runtime/src/qio/qio_formatted.c

```diff
@@ -360,6 +360,68 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
 
   return 0;
 }
+qioerr read_entire_file(qio_channel_t* restrict ch, const char* restrict* restrict out, int64_t* restrict len_out, ssize_t maxlen)
+{
+    qioerr err,err_reading_file;
+    int64_t file_length;
+    err=qio_file_length(ch->file,&file_length); //extracting file size information to allocate buffer with approriate size beforehand
+    char* ret=NULL;
+    ssize_t amt;
+    err_t errcode;
+    if(!err)
+    {
+        
+        //allocate the buffer
+        ret=qio_malloc(file_length+1);
+        ret[0]='\0';
+        err_reading_file=qio_channel_read(false, ch, ret, file_length, &amt);
+        
+    }
+    else           //in case of socket or pipe files, file size info is not readily available.This has also been handled here .
+    {
+        
+        int64_t fileLenGuess=1024;
+        ret=qio_malloc(fileLenGuess+1);
+        while(true)
+        {
+            ret[0]='\0';
+            err_reading_file = qio_channel_read(false, ch, ret, fileLenGuess, &amt);
+            if (err_reading_file == QIO_EEOF) break;
```

Comment:
If we have reached EOF, amt will be some final amount we read. In that case, we should probably reallocate the buffer to only include this much rather than zeroing everything else. Make sure to leave space for the trailing zero byte.

Reply:
Made the changes .